diff --git a/CPP/7zip/7zip_gcc.mak b/CPP/7zip/7zip_gcc.mak
index f86a5fc..ba208dc 100644
--- a/CPP/7zip/7zip_gcc.mak
+++ b/CPP/7zip/7zip_gcc.mak
@@ -23,9 +23,11 @@ PROGPATH = $(O)/$(PROG)
 PROGPATH_STATIC = $(O)/$(PROG)s
 
 
+ifndef EMSCRIPTEN
 ifneq ($(CC), xlc)
 CFLAGS_WARN_WALL = -Werror -Wall -Wextra
 endif
+endif
 
 # for object file
 # -Wa,-aln=test.s
@@ -111,7 +113,11 @@ endif
 
 else
 
+ifdef EMSCRIPTEN
+LDFLAGS = $(LDFLAGS_STATIC) $(LDFLAGS_EMCC)
+else
 LDFLAGS = $(LDFLAGS_STATIC)
+endif
 # -z force-bti
 # -s is not required for clang, do we need it for GCC ???
 
diff --git a/CPP/7zip/Bundles/Alone2/makefile.emcc b/CPP/7zip/Bundles/Alone2/makefile.emcc
new file mode 100644
index 0000000..1e2361c
--- /dev/null
+++ b/CPP/7zip/Bundles/Alone2/makefile.emcc
@@ -0,0 +1,3 @@
+EMSCRIPTEN = 1
+include ./makefile.gcc
+PROG = 7zz.js
diff --git a/CPP/7zip/Bundles/Alone2/post.js b/CPP/7zip/Bundles/Alone2/post.js
new file mode 100644
index 0000000..607ce26
--- /dev/null
+++ b/CPP/7zip/Bundles/Alone2/post.js
@@ -0,0 +1,4 @@
+Module.FS = FS;
+Module.NODEFS = NODEFS;
+Module.WORKERFS = WORKERFS;
+Module.callMain = callMain;
diff --git a/CPP/7zip/Bundles/Alone2/pre.js b/CPP/7zip/Bundles/Alone2/pre.js
new file mode 100644
index 0000000..2084ccb
--- /dev/null
+++ b/CPP/7zip/Bundles/Alone2/pre.js
@@ -0,0 +1 @@
+Module.noInitialRun = true;
diff --git a/CPP/Windows/System.cpp b/CPP/Windows/System.cpp
index ffbf97c..91bf770 100644
--- a/CPP/Windows/System.cpp
+++ b/CPP/Windows/System.cpp
@@ -9,6 +9,7 @@
     || defined(BSD) || defined(__FreeBSD__) || defined(__NetBSD__) || defined(__OpenBSD__) \
     || defined(__QNXNTO__)
 #include <sys/sysctl.h>
+#elif defined(__EMSCRIPTEN__)
 #else
 #include <sys/sysinfo.h>
 #endif
@@ -358,6 +359,7 @@ bool GetRamSize(size_t &size)
   #elif defined(__FreeBSD_kernel__) && defined(__GLIBC__)
   // GNU/kFreeBSD Debian
   // fixme
+  #elif defined(__EMSCRIPTEN__)
   #else
 
   struct sysinfo info;
diff --git a/CPP/Windows/SystemInfo.cpp b/CPP/Windows/SystemInfo.cpp
index a32df74..cbef817 100644
--- a/CPP/Windows/SystemInfo.cpp
+++ b/CPP/Windows/SystemInfo.cpp
@@ -19,7 +19,7 @@
 #ifdef __APPLE__
 #include <sys/sysctl.h>
 
-#elif !defined(_AIX)
+#elif !defined(_AIX) && !defined(__EMSCRIPTEN__)
 
 #if defined(__GLIBC__) && (__GLIBC__ * 100 + __GLIBC_MINOR__ >= 216)
   #define Z7_GETAUXV_AVAILABLE
